src/lib/middleware/ssr.ts

```diff
@@ -0,0 +1,169 @@
+import RSDWClient from "react-server-dom-webpack/client.node.unbundled";
+import { createServer as viteCreateServer } from "vite";
+
+import { configFileConfig, resolveConfig } from "../config.js";
+import { defineEntries } from "../../server.js";
+import type { GetSsrConfig } from "../../server.js";
+import { nonjsResolvePlugin } from "../vite-plugin/nonjs-resolve-plugin.js";
+
+type Middleware = (
+  req: IncomingMessage,
+  res: ServerResponse,
+  next: (err?: unknown) => void
+) => void;
+
+type Entries = { default: ReturnType<typeof defineEntries> };
+
+// eslint-disable-next-line import/no-named-as-default-member
+const { renderToPipeableStream } = RDServer;
+const { createFromNodeStream } = RSDWClient;
+
+const hasStatusCode = (x: unknown): x is { statusCode: number } =>
+  typeof (x as any)?.statusCode === "number";
+
+// TODO make it configurable
+const splitHTML = (htmlStr: string): readonly [string, string] => {
+  const splitted = htmlStr.split(/\s*<div class="spinner"><\/div>\s*/);
+  if (splitted.length !== 2) {
+    throw new Error("Failed to split HTML");
+  }
+  return [splitted[0] as string, splitted[1] as string];
+};
+
+const bundlerConfig = new Proxy(
+  {},
+  {
+    get() {
+      return new Proxy(
+        {},
+        {
+          get() {
+            return { specifier: "waku/server", name: "ClientOnly" };
+          },
+        }
+      );
+    },
+  }
+);
+
+const renderHTML = (
+  pathStr: string,
+  rscServer: URL,
+  ssrConfig: NonNullable<Awaited<ReturnType<GetSsrConfig>>>
+): Readable => {
+  const htmlResPromise = fetch(rscServer + pathStr, {
+    headers: { "x-waku-skip-ssr": "true" },
+  });
+  const [rscId, props] = ssrConfig.element;
+  // FIXME we blindly expect JSON.stringify usage is deterministic
+  const serializedProps = JSON.stringify(props);
+  const searchParams = new URLSearchParams();
+  searchParams.set("props", serializedProps);
+  const rscResPromise = fetch(rscServer + rscId + "/" + searchParams);
+  const passthrough = new PassThrough();
+  Promise.all([htmlResPromise, rscResPromise]).then(
+    async ([htmlRes, rscRes]) => {
+      if (!htmlRes.ok) {
+        const err = new Error("Failed to fetch html from RSC server");
+        (err as any).statusCode = htmlRes.status;
+        passthrough.destroy(err);
+        return;
+      }
+      if (!rscRes.ok) {
+        const err = new Error("Failed to fetch rsc from RSC server");
+        (err as any).statusCode = rscRes.status;
+        passthrough.destroy(err);
+        return;
+      }
+      if (!htmlRes.body || !rscRes.body) {
+        passthrough.destroy(new Error("No body"));
+        return;
+      }
+      const htmlStr = await htmlRes.text(); // Hope stream works, but it'd be too tricky
+      const [preamble, postamble] = splitHTML(htmlStr);
+      passthrough.write(preamble, "utf8");
+      const data = createFromNodeStream(
+        Readable.fromWeb(rscRes.body as any), // FIXME how to avoid any?
```

Comment:
Update to `"@types/node": "^20.3.1"` could fix this issue

This is upstream bug